### PR TITLE
get kubernetes authfication info from kubeConfig

### PR DESCRIPTION
**Pull request by TrsNium**
Please check out the following before check out this PR.(Because it doesn't work without raising the version of kubernetes client)
https://github.com/treasure-data/digdag/pull/1253

This PR  focus on got authentication information of kubernetes.
This PR is also an implementation of the following issue.
https://github.com/treasure-data/digdag/issues/1254

### CODE REVIEW

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 1]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
...
+                 config.has("kube_config_path"))
+            ) {
+            throw new ConfigException("kubernetes config must have master:, certs_ca_data:, oauth_token: and namespace: Or use_kube_config: kube_config_path");
         }
         return config;
     }
 
+    private static io.fabric8.kubernetes.client.Config getKubeConfigFromPath(String path)
+    {
+      try{
```
Comment (sonots) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: nitpick: `try {`

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 2]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
...
+        final Path kubeConfigPath = Paths.get(path);
+        final String kubeConfigContents = new String(Files.readAllBytes(kubeConfigPath), Charset.forName("UTF-8"));
+        return io.fabric8.kubernetes.client.Config.fromKubeconfig(kubeConfigContents);
+      }catch (java.io.IOException e) {
```
Comment (sonots) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: nitpick: `} catch`

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 3]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
...
-                !config.has("namespace")) {
-            throw new ConfigException("kubernetes config must have master:, certs_ca_data:, oauth_token: and namespace:");
+        if (!(config.has("master") &&
+                config.has("certs_ca_data") &&
+                config.has("oauth_token") &&
+                config.has("namespace")) &&
+                !(config.has("use_kube_config") &&
+                 config.has("kube_config_path"))
+            ) {
+            throw new ConfigException("kubernetes config must have master:, certs_ca_data:, oauth_token: and namespace: Or use_kube_config: kube_config_path");
```
Comment (sonots) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: nitpick: `"kubernetes config must have master:, certs_ca_data:, oauth_token: and namespace:, or use_kube_config: 
 and kube_config_path:"`

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 4]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
...
+        if (extracted.get("use_kube_config", boolean.class, false)){
+            io.fabric8.kubernetes.client.Config kubeConfig;
+            kubeConfig = getKubeConfigFromPath(extracted.get("kube_config_path", String.class));
+            return create(clusterName,
+                kubeConfig.getMasterUrl(),
+                kubeConfig.getCaCertData(),
+                kubeConfig.getOauthToken(),
+                kubeConfig.getNamespace()
+              );
+        }else{
```
Comment (sonots) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: nitpick: `} else {`

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 5]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
         }
         final String keyPrefix = KUBERNETES_CLIENT_PARAMS_PREFIX + clusterName + ".";
         final Config extracted = validateParams(StorageManager.extractKeyPrefix(systemConfig, keyPrefix));
-        return create(clusterName,
-                extracted.get("master", String.class),
-                extracted.get("certs_ca_data", String.class),
-                extracted.get("oauth_token", String.class),
-                extracted.get("namespace", String.class));
+        if (extracted.get("use_kube_config", boolean.class, false)){
```
Comment (sonots) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: nitpick: `)) {`

**Comment by muga** [comment]
@TrsNium thank you for improving. i will take a look today.

File: build.gradle [hunk 6]
```diff
@@ -123,8 +123,8 @@ subprojects {
     }
 
     ext {
-        jacksonVersion = "2.9.9"
-        jacksonDatabindVersion = "2.9.9"
+        jacksonVersion = "2.9.10"
+        jacksonDatabindVersion = "2.9.10"
```
Comment (muga) [inline build.gradle]: same as https://github.com/treasure-data/digdag/pull/1253? the PR was merged into v0_10.

File: digdag-standards/build.gradle [hunk 7]
```diff
@@ -41,7 +41,13 @@ dependencies {
...
 
     // kubernetes
-    compile ('io.fabric8:kubernetes-client:3.1.5')
+    compile ('io.fabric8:kubernetes-client:4.6.2') {
+        // Avoid compilation failures due to dependencies
+        exclude group: 'org.slf4j', module: 'slf4j-api'
+        // TODO: digdag use version 3.12.0, upgrade to version 3.12.6.
+        exclude group: 'com.squareup.okhttp3', module: 'okhttp'
+        exclude group: 'com.fasterxml.jackson.dataformat', module: 'jackson-module-jaxb-annotations'
+    }
```
Comment (muga) [inline digdag-standards/build.gradle]: ditto

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 8]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
         }
         final String keyPrefix = KUBERNETES_CLIENT_PARAMS_PREFIX + clusterName + ".";
         final Config extracted = validateParams(StorageManager.extractKeyPrefix(systemConfig, keyPrefix));
-        return create(clusterName,
-                extracted.get("master", String.class),
-                extracted.get("certs_ca_data", String.class),
-                extracted.get("oauth_token", String.class),
-                extracted.get("namespace", String.class));
+        if (extracted.get("use_kube_config", boolean.class, false)) {
```
Comment (muga) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: I'm wondering that 'use_kube_config' key is actually necessary or not. instead 'kube_config_path' key could be used for this condition?

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java [hunk 9]
```diff
@@ -46,24 +51,49 @@ private static KubernetesClientConfig createFromSystemConfig(final Optional<Stri
...
+            io.fabric8.kubernetes.client.Config kubeConfig;
+            kubeConfig = getKubeConfigFromPath(extracted.get("kube_config_path", String.class));
```
Comment (muga) [inline digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: What do you think that 'validateParams' could be called for this 'kubeConfig' params? It means the impl. of the existing 'validateParams' could not be changed.

**Comment by TrsNium** [comment]
@muga 
hank you for the review! As it makes sense, I will fix that part.
I have one question. I want to put a KubeConfig for testing. Where should I put it?

**Comment by muga** [comment]
> I want to put a KubeConfig for testing. Where should I put it?

@TrsNium Thanks for taking it! appreciated. according to naming convention, the test will be in digdag-standards/src/test/java/io/digdag/standards/command/kubernetes/KubernetesClientConfigTest.java. if you have any questions, please let us know.

**Comment by TrsNium** [comment]
I found that getOauthToken and getCaCertData return null, so I'll fix it

**Comment by TrsNium** [comment]
@muga Thank you for reply!
I Added tests and partially modified code. Please check them.

File: digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java (hunk 9 above)
Comment (TrsNium) [reply digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: Is it OK to "Check the value of` kubeConfig’s parameters` are null "?

Comment (muga) [reply digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: good point. they should be validated. up to you but one idea is to add new 'validateKubeConfig' method in this class. in this case,  we'd better to rename the existing 'validateParams' with 'validateConfig'. what do you think?

Comment (TrsNium) [reply digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java]: > we'd better to rename the existing 'validateParams' with 'validateConfig'
I think so too, rename `validateParams` to `validateConfig`.

I also add `validateKubeConfig` method for kubeConfig’s.

**Comment by TrsNium** [comment]
@muga 
Thank you for review!
Update access level of method tested and add a method to validate KubeConfig.
Please check them.

**Comment by yoyama** [comment]
@TrsNium Merged. Thanks for your contribution.

**Pull request merged**